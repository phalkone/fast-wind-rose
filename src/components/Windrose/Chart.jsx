/* eslint-disable react/jsx-fragments */
import React, { Fragment } from 'react'
import PropTypes from 'prop-types'

/**
 * The chart that can be divided into a specified number of sectors
 */
function Chart (props) {
  return (
    <Fragment>
      {/* Sector lines */}
      {[...new Array(props.sectorCount).keys()].map(i => (
        <Fragment key={i}>
          <line
            x1={props.center}
            x2={props.center}
            y1={10}
            y2={props.center}
            stroke='grey'
            strokeWidth='1'
            transform={`rotate(${(0.5 + i) * (360 / props.sectorCount)},` +
                       ` ${props.center},${props.center})`}
          />
        </Fragment>
      ))}
      {/* Circles */}
      {[...new Array(4).keys()].map(i => (
        <Fragment key={i}>
          <circle
            cx={props.center}
            cy={props.center}
            r={(4 - i) * ((props.center - 10) / 4)}
            fill='none'
            strokeWidth='1'
            stroke='grey'
            opacity={i > 0 ? 0.5 : 1}
          />
        </Fragment>
      ))}
    </Fragment>
  )
}

Chart.propTypes = {
  /**
   * The number of sectors
   */
  sectorCount: PropTypes.number,
  /**
   * The center of the chart. x and y coordinates are the same.
   */
  center: PropTypes.number
}

export default Chart
