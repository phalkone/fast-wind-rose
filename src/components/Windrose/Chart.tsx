import React, { Fragment, useContext } from 'react'
import { WindroseContext } from '.'

/**
 * The chart that can be divided into a specified number of sectors
 */
export const Chart = () => {
  const context = useContext(WindroseContext)
  return (
    <>
      {/* Sector lines */}
      {[...new Array(Math.round(360 / context.sectorSize)).keys()].map(i => (
        <Fragment key={i}>
          <line
            x1={context.center}
            x2={context.center}
            y1={10}
            y2={context.center}
            stroke='grey'
            strokeWidth='1'
            transform={`rotate(${(0.5 + i) * (context.sectorSize)},` +
                       `${context.center},${context.center})`}
          />
        </Fragment>
      ))}
      {/* Circles */}
      {[...new Array(4).keys()].map(i => (
        <Fragment key={i}>
          <circle
            cx={context.center}
            cy={context.center}
            r={(4 - i) * ((context.center - 10) / 4)}
            fill='none'
            strokeWidth='1'
            stroke='grey'
            opacity={i > 0 ? 0.5 : 1}
          />
        </Fragment>
      ))}
    </>
  )
}
