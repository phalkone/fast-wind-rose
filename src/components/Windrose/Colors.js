const colors = {
  0: 'rgb(60,95,156)',
  5: 'rgb(94,131,188)',
  10: 'rgb(143,180,232)',
  15: 'rgb(174,203,214)',
  20: 'rgb(220,226,220)',
  25: 'rgb(251,233,94)',
  30: 'rgb(252,195,67)',
  35: 'rgb(245,136,42)',
  40: 'rgb(242,103,33)',
  45: 'rgb(243,63,29)',
  50: 'rgb(244,36,27)'
}

export default colors
