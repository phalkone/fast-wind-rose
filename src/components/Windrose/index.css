button {
  position: absolute;
  top: 10px;
  color: rgb(0, 0, 0);
  background-color: white;
  font-weight: bolder;
  border: 1px solid black;
  text-decoration: none;
  padding: 0 5px;
  transition-duration: 0.4s;
}

button:hover {
  background-color: rgb(203, 203, 203);
}

select {
  position: absolute;
  top: 10px;
  left: 10px;
  padding: 2px;
}

svg, select, button {
  font-family: Roboto, "Open Sans", sans-serif;
}
