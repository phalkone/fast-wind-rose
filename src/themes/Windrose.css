.windrose-button{
  position: absolute;
  top: 10px;
  color: rgb(0, 0, 0);
  background-color: white;
  font-weight: bolder;
  border: 1px solid black;
  text-decoration: none;
  padding: 0 5px;
  transition-duration: 0.4s;
}
 
.windrose-button:hover {
  background-color: rgb(203, 203, 203);
}

.windrose-select {
  position: absolute;
  top: 10px;
  left: 10px;
  padding: 2px;
}

.windrose-container {
  font-family: Roboto, "Open Sans", sans-serif;
}