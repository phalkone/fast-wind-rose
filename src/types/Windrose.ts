export interface IWindrose {
  /**
   * Hide or display legend
   */
  legend: boolean,
  /**
   * Width/height of chart. Will be displayed in specified viewbox.
   */
  size: number,
  /**
   * Default number of sectors. Must be included in sectorArray.
   */
  sectorCount: number,
  /**
   * Fixed interval between data points specified in hours. For example if
   * data points are every 30 min, then value should be 0.5
   */
  interval: number,
  /**
   * Scale of speeds with the linked color. Example as follows:
   *  { 0: 'rgb(60,95,156)', 5: 'rgb(94,131,188)' }
   */
  scale: { [n: number]: string },
  /**
   * Array with directional data for the wind
   */
  dirData: object[],
  /**
  * Array with speed data for the wind
  */
  spdData: object[],
  /**
   * Key for speed data
   */
  spdKey: string,
  /**
   * Key for directional data
   */
  dirKey: string,
  /**
   * Scales the compilation scale against the viewbox. Choose a value lower
   * than 1 to scale down and larger than 1 to scale up.
   */
  enlarge: number,
  /**
   * Array that defines the number of sectors the user can choose from.
   */
  sectorArray: number[],
  /**
   * Common key between direction and speed data
   */
  commonKey: string
}

export interface ILegend {
  /**
   * Size of the chart. Width = Height.
   */
  size: number,
  /**
   * Scale of speeds with the linked color. Example as follows:
   *  { 0: 'rgb(60,95,156)', 5: 'rgb(94,131,188)' }
   */
  scale: { [n: number]: string }
}

export interface IShip {
  /**
   * The center of the ship path. x and y coordinates are assumed to be the same.
   */
  center: number
}

export interface IChart extends IShip {
  /**
   * The number of sectors
   */
  sectorSize: number
}

export interface IIntervalLabel extends IChart {
  /**
   * Defines which sector is being drawn. 0 being the first sector.
   */
  sector: number,
  /**
   * Interval between data points specified in hours
   */
  interval: number,
  /**
   * Array with the speeds that are part of this sector
   */
  speeds: number[]
}

export interface ISector extends IIntervalLabel, ILegend {
  /**
   * Total bar length of this sector. The sector with the largest interval will
   * have a bar length equal to the radius.
   */
  barLength: number,
  /**
   * Defines the bar length of 1 unit of interval.
   */
  unit: number,
}

export interface IToolTip {
  /**
   * x coordinate of top left corner of tooltip
   */
  x: number,
  /**
   * y coordinate of top left corner of tooltip
   */
  y: number,
  /**
   * text of tooltip
   */
  text: [string, number],
  /**
   * Scale factor due to viewBox of SVG vs actual size of SVG.
   */
  factor: number
}
