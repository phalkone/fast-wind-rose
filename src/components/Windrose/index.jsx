import React, { Fragment } from 'react'
import Legend from './Legend'
import Chart from './Chart'
import Ship from './Ship'
import Sector from './Sector'
import IntervalLabel from './IntervalLabel'
import PropTypes from 'prop-types'

export default class Windrose extends React.Component {
  constructor (props) {
    super(props)
    const sectors = [...new Array(360 / this.props.sectorSize).keys()].map((el) => [])
    for (let i = 0; i < this.props.data.length; i++) {
      let dir = this.props.data[i].direction + (this.props.sectorSize / 2)
      if (dir > 360) dir -= 360
      const cat = Math.floor(dir / this.props.sectorSize)
      sectors[cat].push(this.props.data[i].speed)
    }
    this.max = sectors.map((el) => el.length).sort((a, b) => b - a)[0]
    this.sectors = sectors
  }

  renderSectors () {
    return (
      this.sectors.map((speeds, i) => (
        <Fragment key={i}>
          {speeds.length &&
            <IntervalLabel
              sector={i}
              interval={speeds.length}
              sectorSize={this.props.sectorSize}
            />}
          {speeds.length &&
            <Sector
              sector={i}
              speeds={speeds}
              sectorSize={this.props.sectorSize}
              barLength={(speeds.length / this.max) * 120}
              interval={120 / this.max}
            />}
        </Fragment>
      ))
    )
  }

  render () {
    return (
      <svg
        version='1.1'
        viewBox='0 0 325 260'
        width={this.props.width}
        height={this.props.height}
      >
        {this.props.legend && <Legend />}
        <Chart sectorSize={this.props.sectorSize} />
        {this.renderSectors()}
        <Ship />
      </svg>
    )
  }
}

Windrose.propTypes = {
  height: PropTypes.number,
  width: PropTypes.number,
  sectorSize: PropTypes.number,
  legend: PropTypes.bool,
  data: PropTypes.array
}
