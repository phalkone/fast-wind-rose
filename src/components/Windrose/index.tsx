import React, { createContext, Fragment, useState } from 'react'
import { Legend } from './Legend'
import { Chart } from './Chart'
import { Ship } from './Ship'
import { Sector } from './Sector'
import { divideBySector } from '../../utils/Windrose.util'
import type { IWindrose, IWindroseContext } from '../../types/Windrose.types'
import '../../themes/Windrose.scss'

/**
 * Draws a windrose for provided relative wind data
 * @param {IWindrose} props Properties of the Windrose as descibed by IWindrose.
 */
const Windrose = (props: IWindrose) => {
  /* Use state to set the number of sectors and show/hide legend */
  const [sectorCount, setSectorCount] = useState<number>(props.sectorCount)
  const [legend, setLegend] = useState<boolean>(props.legend)

  /* Divide data points per sector */
  const sectors = divideBySector(sectorCount, props.dirData, props.dirKey,
    props.spdData, props.spdKey, props.commonKey)

  /* Set the viewBox size compared to the compilation size */
  const compilationSize = 260 / props.enlarge

  /* Add sectorCount if not part of sectorArray */
  if (!props.sectorArray.includes(sectorCount)) {
    props.sectorArray.push(sectorCount)
  }

  return (
    <div className='windrose-container'>
      {/* Selector for the number of sectors */}
      <select
        title='Number of sectors'
        value={sectorCount}
        className='windrose-select'
        onChange={e => { setSectorCount(Number(e.target.value)) }}
      >
        {props.sectorArray.map(s => (
          <Fragment key={s}>
            <option value={s}>{s}</option>
          </Fragment>
        ))}
      </select>
      <button
        title='Show/hide legend'
        className='windrose-button'
        onClick={e => { setLegend(!legend) }}
        style={{
          left: props.size - 30
        }}
      >{legend ? '<<' : '>>'}
      </button>
      {/* Windrose SVG with optional legend */}
      <svg
        version='1.1'
        viewBox={`0 0 ${compilationSize + (legend ? 65 : 0)} ${compilationSize}`}
        width={props.size + (legend ? 65 * (props.size / compilationSize) : 0)}
        height={props.size}
      >
        <WindroseContext.Provider value={{
          center: compilationSize / 2,
          scale: props.scale,
          sectorSize: 360 / sectorCount,
          interval: props.interval,
          size: props.size,
          max: sectors.map((el) => el.length).sort((a, b) => b - a)[0]
        }}>
          {/* Draw legend only if option is given as prop */}
          {legend && <Legend/>}
          <Chart/>
          {/* Draw each sector with interval label */}
          {sectors.map((speeds, i) => (
            speeds.length && <Fragment key={i}>
              <Sector
                sector={i}
                speeds={speeds}
              />
            </Fragment>
          ))}
          {/* Draw ship outline */}
          <Ship/>
        </WindroseContext.Provider>
        {/* Ensures that tooltips are always on top */}
        <use href='#tooltip' fill='black' />
        <use href='#tooltiptext' />
      </svg>
    </div>
  )
}

Windrose.defaultProps = {
  size: 260,
  spdKey: 'value',
  dirKey: 'value',
  commonKey: 'timestamp',
  sectorCount: 12,
  enlarge: 1,
  interval: 1,
  sectorArray: [4, 8, 12, 16, 24, 32, 36],
  scale: {
    0: 'rgb(60,95,156)',
    5: 'rgb(94,131,188)',
    10: 'rgb(143,180,232)',
    15: 'rgb(174,203,214)',
    20: 'rgb(220,226,220)',
    25: 'rgb(251,233,94)',
    30: 'rgb(252,195,67)',
    35: 'rgb(245,136,42)',
    40: 'rgb(242,103,33)',
    45: 'rgb(243,63,29)',
    50: 'rgb(244,36,27)'
  }
}

const defaultContext : IWindroseContext = {
  center: 130,
  scale: Windrose.defaultProps.scale,
  sectorSize: 30,
  interval: 1,
  size: 260,
  max: 24
}

export const WindroseContext = createContext(defaultContext)

export default Windrose
