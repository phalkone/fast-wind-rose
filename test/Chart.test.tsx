import React from 'react'
import { render, unmountComponentAtNode } from 'react-dom'
import { act } from 'react-dom/test-utils'
import { Chart } from '../src/components/Windrose/Chart'
import { WindroseContext } from '../src/components/Windrose'
import renderer from 'react-test-renderer'

let container : HTMLElement
beforeEach(() => {
  container = document.createElement('svg')
  container.setAttribute('viewBox', '0 0 325 260')
  document.body.appendChild(container)
})

afterEach(() => {
  unmountComponentAtNode(container)
  container.remove()
})

it('renders the correct number of lines', () => {
  act(() => {
    render(
      <WindroseContext.Provider value={{ sectorSize: 10, center: 130 }}>
        <Chart/>
      </WindroseContext.Provider>
      , container)
  })
  expect(Array.from(container.children).filter((el : Element) => el.tagName === 'LINE').length).toEqual(36)
})

it('renders the correct number of circles', () => {
  act(() => {
    render(
      <WindroseContext.Provider value={{ sectorSize: 10, center: 130 }}>
        <Chart/>
      </WindroseContext.Provider>
      , container)
  })
  expect(Array.from(container.children).filter((el : Element) => el.tagName === 'CIRCLE').length).toEqual(4)
})

it('renders lines and circles with the correct center', () => {
  act(() => {
    render(
      <WindroseContext.Provider value={{ sectorSize: 30, center: 130 }}>
        <Chart/>
      </WindroseContext.Provider>
      , container)
  })
  const centers = Array.from(container.children)
    .filter((el : Element) => el.tagName === 'CIRCLE')
    .map((el: Element) => [el.getAttribute('cx'), el.getAttribute('cy')])
    .every((el: any) => el[0] === '130' && el[1] === '130')
  const lines = Array.from(container.children)
    .filter((el : Element) => el.tagName === 'LINE')
    .map((el: Element) => [el.getAttribute('x1'), el.getAttribute('x2'), el.getAttribute('y2')])
    .every((el: any) => el[0] === '130' && el[1] === '130' && el[2] === '130')
  expect(centers && lines).toBeTruthy()
})

it('renders circles with the correct diameter', () => {
  act(() => {
    render(
      <WindroseContext.Provider value={{ sectorSize: 30, center: 130 }}>
        <Chart/>
      </WindroseContext.Provider>
      , container)
  })
  const radius = Array.from(container.children)
    .filter((el : Element) => el.tagName === 'CIRCLE')
    .map((el: Element) => el.getAttribute('r'))
  expect(radius).toEqual(['120', '90', '60', '30'])
})

it('renders lines with the correct rotation', () => {
  act(() => {
    render(
      <WindroseContext.Provider value={{ sectorSize: 90, center: 130 }}>
        <Chart/>
      </WindroseContext.Provider>
      , container)
  })
  const radius = Array.from(container.children)
    .filter((el : Element) => el.tagName === 'LINE')
    .map((el: Element) => el.getAttribute('transform'))
  expect(radius).toEqual(['rotate(45, 130,130)', 'rotate(135, 130,130)',
    'rotate(225, 130,130)', 'rotate(315, 130,130)'])
})

test('render a chart', () => {
  const comp = renderer.create(
    <WindroseContext.Provider value={{ sectorSize: 30, center: 130 }}>
      <Chart/>
    </WindroseContext.Provider>
  )
  const tree = comp.toJSON()
  expect(tree).toMatchSnapshot()
})
