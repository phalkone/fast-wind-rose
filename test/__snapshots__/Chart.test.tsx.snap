// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render a chart 1`] = `
Array [
  <line
    stroke="grey"
    strokeWidth="1"
    transform="rotate(15,130,130)"
    x1={130}
    x2={130}
    y1={10}
    y2={130}
  />,
  <line
    stroke="grey"
    strokeWidth="1"
    transform="rotate(45,130,130)"
    x1={130}
    x2={130}
    y1={10}
    y2={130}
  />,
  <line
    stroke="grey"
    strokeWidth="1"
    transform="rotate(75,130,130)"
    x1={130}
    x2={130}
    y1={10}
    y2={130}
  />,
  <line
    stroke="grey"
    strokeWidth="1"
    transform="rotate(105,130,130)"
    x1={130}
    x2={130}
    y1={10}
    y2={130}
  />,
  <line
    stroke="grey"
    strokeWidth="1"
    transform="rotate(135,130,130)"
    x1={130}
    x2={130}
    y1={10}
    y2={130}
  />,
  <line
    stroke="grey"
    strokeWidth="1"
    transform="rotate(165,130,130)"
    x1={130}
    x2={130}
    y1={10}
    y2={130}
  />,
  <line
    stroke="grey"
    strokeWidth="1"
    transform="rotate(195,130,130)"
    x1={130}
    x2={130}
    y1={10}
    y2={130}
  />,
  <line
    stroke="grey"
    strokeWidth="1"
    transform="rotate(225,130,130)"
    x1={130}
    x2={130}
    y1={10}
    y2={130}
  />,
  <line
    stroke="grey"
    strokeWidth="1"
    transform="rotate(255,130,130)"
    x1={130}
    x2={130}
    y1={10}
    y2={130}
  />,
  <line
    stroke="grey"
    strokeWidth="1"
    transform="rotate(285,130,130)"
    x1={130}
    x2={130}
    y1={10}
    y2={130}
  />,
  <line
    stroke="grey"
    strokeWidth="1"
    transform="rotate(315,130,130)"
    x1={130}
    x2={130}
    y1={10}
    y2={130}
  />,
  <line
    stroke="grey"
    strokeWidth="1"
    transform="rotate(345,130,130)"
    x1={130}
    x2={130}
    y1={10}
    y2={130}
  />,
  <circle
    cx={130}
    cy={130}
    fill="none"
    opacity={1}
    r={120}
    stroke="grey"
    strokeWidth="1"
  />,
  <circle
    cx={130}
    cy={130}
    fill="none"
    opacity={0.5}
    r={90}
    stroke="grey"
    strokeWidth="1"
  />,
  <circle
    cx={130}
    cy={130}
    fill="none"
    opacity={0.5}
    r={60}
    stroke="grey"
    strokeWidth="1"
  />,
  <circle
    cx={130}
    cy={130}
    fill="none"
    opacity={0.5}
    r={30}
    stroke="grey"
    strokeWidth="1"
  />,
]
`;
